### PR TITLE
Move BinaryReader to shared CoreLib partition

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

File: src/System.Private.CoreLib/src/System/Decimal.cs [hunk 1]
```diff
@@ -156,19 +156,16 @@ public partial struct Decimal : IFormattable, IComparable, IConvertible, ICompar
         //
         public Decimal(int value)
         {
-            //  JIT today can't inline methods that contains "starg" opcode.
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Decimal.cs]: This bug was fixed in CoreCLR a long time ago (#6572)

File: src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs [hunk 2]
```diff
@@ -220,8 +216,7 @@ public virtual int Read()
 
                 Debug.Assert(charsRead < 2, "BinaryReader::ReadOneChar - assuming we only got 0 or 1 char, not 2!");
             }
-            if (charsRead == 0)
-                return -1;
+            Debug.Assert(charsRead > 0);
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs]: Unrechable code (cleanup from CoreRT)

File: src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs [hunk 3]
```diff
@@ -484,15 +479,10 @@ private int InternalReadChars(Span<char> buffer)
                 // do ~1+log(n) reads to read n characters.
                 numBytes = charsRemaining;
 
-                // special case for DecoderNLS subclasses when there is a hanging byte from the previous loop
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs]: This hack should not be needed (cleanup from CoreRT). I was not able to trace down why it was added.

File: src/System.Private.CoreLib/src/System/Decimal.cs [hunk 4]
```diff
@@ -266,15 +260,6 @@ public static decimal FromOACurrency(long cy)
...
         public Decimal(int[] bits)
-        {
-            lo = 0;
-            mid = 0;
-            hi = 0;
-            flags = 0;
-            SetBits(bits);
-        }
-
-        private void SetBits(int[] bits)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Decimal.cs]: Simplified this a bit while I was on it.

File: src/System.Private.CoreLib/src/System/Decimal.cs [hunk 5]
```diff
@@ -313,7 +298,7 @@ private void OnSerializing(StreamingContext ctx)
             // OnSerializing is called before serialization of an object
             try
             {
-                SetBits(GetBits(this));
+                new decimal(lo, mid, hi, flags);
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/System/Decimal.cs]: I think I am missing something obvious here. How does just creating a new decimal instance address anything? The previous code seems to have done a get/set of values on `this` instance where as now we create a new instance that is just collected - isn't this just a no-op?

Comment (AaronRobinsonMSFT) [reply src/System.Private.CoreLib/src/System/Decimal.cs]: Oh... i see now. Wow this is rather convoluted. Would it be possible to just have a simple `ValidateParts()` static function that either returned `true`/`false` which then we could throw a specific exception? The private constructor just for validation seems weird.

Comment (AaronRobinsonMSFT) [reply src/System.Private.CoreLib/src/System/Decimal.cs]: Sad 😞 Okay, well thanks for considering. Can you add an explicit comment in both locations indicating the reason they type is being created is to validate the arguments?

**Comment by jkotas** [comment]
@dotnet-bot test CentOS7.1 x64 Checked Innerloop Build and Test
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests please

**Comment by jkotas** [comment]
@Anipik The mirror is not picking up this change. Could you please take a look?

**Comment by Anipik** [comment]
@jkotas done

**Pull request merged**